src/components/05_pages/Permissions/Permissions.js

```diff
@@ -26,67 +28,48 @@ export const filterPermissions = (input, permissions) =>
 let styles;
 
 const Permissions = class Permissions extends Component {
-  static propTypes = {
-    setMessage: func.isRequired,
-    clearMessage: func.isRequired,
-    match: shape({
-      params: shape({
-        role: string,
-      }).isRequired,
-    }).isRequired,
-  };
   state = {
     loaded: false,
-    rawPermissions: [],
-    renderablePermissions: [],
+    permissions: [],
+    changedRoles: [],
+    filter: '',
     working: false,
   };
   componentDidMount() {
-    this.cancelFetch = this.fetchData();
+    this.props.requestPermissions();
+    this.props.requestRoles();
   }
   componentWillUnmount() {
-    this.cancelFetch();
+    this.props.cancelTask();
   }
   onPermissionCheck = (roleName, permission) => {
     this.setState(prevState => ({
-      changedRoles: [...new Set(prevState.changedRoles).add(roleName).values()],
+      ...prevState,
+      changedRoles: Array.from(
+        new Set(prevState.changedRoles).add(roleName).values(),
+      ),
       roles: this.togglePermission(permission, roleName, prevState.roles),
```

Comment:
`prevState.roles` does not exist until you make the first change to permissions.

Reply:
good catch!